### ISSUE TITLE
Right click menu and tab icon

### ISSUE THREAD

**Issue by Jollfye** [deanoemcke/thegreatsuspender#808]
Just noticed another two not so annoying "bugs".

1. When I go to right click TGS menu it looks too big, maybe it's possible to halve it?

![scr](https://user-images.githubusercontent.com/6180818/47899322-9d45f800-de92-11e8-9e66-f2418165d78b.png)

2. Tab's icon not shows up when suspended, just blank, even if when tab's unsuspended its icon is there (as it should be). Opening another or several tabs with the same domain doesn't solve the problem, coz they all getting blank icon when suspended. Tho for now I have that issue with just one tab, still I decided to write bout it here. + idk how or coz of what this problem started, it's just I noticed icon not being there.

![scr1](https://user-images.githubusercontent.com/6180818/47900143-4b52a180-de95-11e8-9324-79f44acc65c3.png)

### Specifications:
 * Extension version: 7.0.140
 * Browser & version: Google Chrome 72.0.3595.2 (Official Build) dev (64-bit)
 * Operating system & version: Windows 10 1809 17763.55

**Comment by Jollfye** [comment]
Regarding the second issue, tab in question close, chrome full restart, open tab/url again, it helped. Now when suspended, icon is there, no more blank. idk what it was.

Plus two more things.

3. When chrome starts (uses CPU 100% n memory like crazy, not touching it) it loads (not unsuspends I guess) all suspended tabs, while so, some tabs' icons becoming or being not pale (like in not suspended state), while others being pale (as they should), 50/50 I think, some bright, some pale, then with time when chrome finishes (CPU/RAM (more than 2/3)/disk load drops) and ready to work, most of the tabs (almost all) getting paled/shadowed, but there's some 5-10 tabs still being left bright, and when I go clicking through them, they being like "loaded" and their icons paled. Not a big issue for now, I think.

4. When exporting/saving session to txt it doesn't include titles (checked with a new 140, same)? icons? So when I imported, then resuspended this session (when I uninstalled TGS store 109 version to install unpacked 131 then, I had not so old session txt 109 backup before chrome got updated and session management stopped working completely), well icons were there, but no titles, just urls, hmm.. Is it possible to save/load urls' titles/positions when exporting/importing session? + It's good that extension folder name and suspended.html url not being saved to txt (I'm talking bout "chrome-extension://oojijclmpkicnfdlc..blabla/suspended.html"), coz they r different with store and unpacked version. + Also thought bout exporting TGS extension settings, like options, shortcuts, whitelisted sites, etc.. that would be nice feature, but yeah not so necessary when u not uninstalling extension/changing folders, but there could be use cases, like when using ur settings on a new PC/OS without Google Chrome account login, just saying, for the future maybe.

Alright, that's all for now, I think. Thanks.

**Comment by Jollfye** [comment]
Restarted PC several times recently, even if all chrome windows closed (last closed window had all tabs suspended) before restart (but it still works in background/sits in tray), when Windows restarts at start it still "restores" chrome with TGS "restoring" my last closed window/session and doubling it, soo yep, issue is still here, 7.0.140. I guess my previous corresponding feedback was incorrect.

**Comment by deanoemcke** [comment]
@Jollfye Thanks for the bugs reports. There's a lot here :/ 
I'll address them in order:

1. Have you tried the most recent version (v7.0.142)? It's possible that this has been fixed. On my dev version it looks like this:
![image](https://user-images.githubusercontent.com/1051763/48450228-5816b400-e80a-11e8-82be-a6cc93be970c.png)

2. Once again, if you could try on v7.0.142 this issue may have been resolved. I've just finished refactoring all the code that handles the favicons.

3. The high CPU load is unavoidable unfortunately. in fact, the only reason this is necessary is to ensure the correct favicons get set. The latest version has made some slight optimisations in this process, but the 'bright' favicon is a fallback while waiting for the faded favicon to be produced. The fact that sometimes some favicons do not get replaced suggests that they are failing to completely load. Not sure why this might be.

4. For the exporting, I'm aware of the limitations of the current export. For now the idea is that it's primarily for exporting to some other application, so i wanted to keep it generic and simple - hence the newline separated unsuspended urls. I agree there's definitely room in the future for providing some other kind of export that includes more metadata such as tab title, favicon, and scroll position. This could perhaps also include settings etc and allow for backing up all sessions in one go. If you want to create a separate feature request issue for this on GitHub you're welcome too :)

5. (Your final comment about doubling of tabs on startup), this is a troubling issue. I'd love to find out more information about it from you if possible. 
What do you have set in chrome -> settings -> On start-up? 
Can you repeat the issue consistently? If so, can you please give me the exact steps to reproduce?

Thanks for all your feedback!

**Comment by deanoemcke** [comment]
@Jollfye for the last issue (no. 5) it is perhaps a duplicate of issue https://github.com/deanoemcke/thegreatsuspender/issues/781 which should now be resolved in the latest version (v7.0.143)

**Comment by Jollfye** [comment]
@deanoemcke Installed latest 143 version + Chrome got updated to 72.0.3608.4 (Official Build) dev (64-bit):
1. Well, unfortunately it's not fixed, looks the same as on my above screeny.

2. Tried, the problem appeared again, all YouTube tabs were blank after I restored session (window) using TabCloud extension (coz it saves page's titles when un/installing TGS extension, tho now I just copy-paste files in extension folder and let it update itself on start rather than un/install it manually). I prefer externally save (or restore, if smth wrong) windows coz there was time TGS Session management didn't work after yet another Chrome update. Tho the blank favicons issue was solved when I completely (with exit) restarted Google Chrome, all tabs' icons were reloaded, good.

3. I think high CPU load at start is ok, so that now I know for sure where it comes from. About the issue, I tried full restart Chrome several times already and for now almost all tabs' icons being faded when start process finishes and load drops. Almost, so as u said, some tabs are failing to completely load (some have blank, others TGS extension icon instead of their favs), for me it's 5-7 of them (not much), but It's not a big deal, coz just click/load them is enough for favicon to appear + I don't see bright icons anymore, that's good, they all TGS icon at start, then CPU load, then yeah, all as described above.

4. Mm that backup feature would be nice, for in case smth goes wrong with extension/Chrome updates or for other use cases (like extension folder name change when switching from store version to unpacked, etc.). So thanks, maybe later I create a separate request. For now I'm ok with external (TabCloud) tabs/windows saving (just in case) and copy-paste files from zip TGS extension updating. Definitely not urgent thing, soo for later.

5. Yep, I guess it was a duplicate, but now, with 143, issue is finally solved! No more on Windows startup Chrome "restoring" last window and doubling it's tabs + no more doubling on each Chrome full restart + Session management always works and exports tabs to txt correctly. Also I noticed Chrome now uses even less memory when all tabs suspended than before TGS versions (with "built-in memory-saving" checked). Great! 👍

**Comment by deanoemcke** [comment]
@Jollfye 
1. This should be fixed in the latest build. It actually means on my computer it looks like this: 
![image](https://user-images.githubusercontent.com/1051763/48741302-77f12080-ecbf-11e8-9173-ed85f2f7d280.png)
But it's a trade off and should look normal on yours (and I think most others).

2. The latest version also contains some more favicon fixes. Perhaps this will finally resolve your favicon issues?

3. Once again, the favicon fixes in the latest release will hopefully resolve any outstanding issues you have with favicons not being set properly after chrome restart.

Please let me know if the latest version resolves some of your outstanding issues.

**Comment by Jollfye** [comment]
@deanoemcke Installed 144 + Google Chrome now is 72.0.3610.2 (Official Build) dev (64-bit):

1. Yep, fixed, for me it looks like this, all good and fine:

![scr](https://user-images.githubusercontent.com/6180818/48779992-434a8c80-ecf2-11e8-920e-1ef435a50377.jpg)

2. Well there's no more completely all blank icons for one site (like YouTube or else), now it's more problem 3.

3. a) I restarted Chrome to update TGS and when it started there were left unfaded 3 bright icons + 4 YouTube blank icons (they were like before: "click/load them is enough for favicon to appear" and be faded) but no more TGS extension like icons were present after load. + When Chrome was starting first I saw all tabs' icons being TGS extension like then I saw them all bright for less than a second and then they all faded (I guess that's normal), good, juust except those 7. b) Then I full restarted Chrome again and there was a messs, half tabs were left bright, some blank, some TGS like, so I went through all of them with ctrl + tab and they all loaded/faded as usual, hmm. c) After, I restarted Chrome for the third time, aand when it loaded there were no bright, blank or TGS like icons left, just all faded, hmmm. d) Restarted again, TGS like icons -> bright -> all faded, hmmmm. I think it might be fixed for now, maybe all icons now stored or smth idk the "new way" with some your new fixes so they don't appear else than faded. Soo I'll let u know here if issue appears again.

**Comment by deanoemcke** [comment]
@Jollfye Sorry, but I've just made another refactor of the favicon code.
I managed to replicate your issue and made some changes to clean up the code and improve startup performance.
Want to give it another shot? v7.0.145

There should be no intermediate blank or bright icons anymore. Suspended tabs should go straight from the extension (suspendy face) icon to the correct faded favicon.

**Comment by Jollfye** [comment]
Updated TGS to 147, Google Chrome is the same 72.0.3610.2 (Official Build) dev (64-bit), full restarted Chrome couple times aand looks like it's all fixed now, as u said no more blank or bright icons, at start I see how suspended tabs (suspendy face) gradually loading (from left to right, all of them, smth like 5 tabs at a time) and being faded every time I restart Chrome (no else than faded icons being left after), Great.

Just one minor thing I noticed while updating TGS and basically I think it's OK. Here's what I did:

1. All tabs suspended. Somewhere in the middle of all tabs one tab selected, YouTube for example.
2. Chrome -> Exit.
3. Copy paste new files from zip to unpacked TGS extension folder.
4. Chrome start.

And here's the thing. When Chrome starts with TGS ext files updated this "updated" page opens up. So here Chrome starts with all tabs suspendy face and starts loading them, but there's a moment when tabs' loading was just started and here opens "updated" page changing focus from previously (before Chrome exit/TGS files update) selected tab (somewhere in the middle) to this "updated" page (at the end) showing it, at this time I saw previously selected tab being bright while all else being suspendy face to faded loading, when loading got to this bright tab it was faded as all others before and after it, so not an issue, just decided to mention it here (if we're talking no blank or bright tabs' icons like at all). + When I just restart Chrome (no TGS update) I don't see any blank or bright icons (only suspendy to faded) + I don't see current or last selected (before exit) tab being bright coz it's focused and has a close cross X on it instead any icon and I don't try to change tabs while Chrome starts/loading. So here I guess the bright icon now appears only once with TGS extension update and "updated" page focus.

**Comment by deanoemcke** [comment]
@Jollfye Thanks for getting back to me with your findings. Good to hear it's _mostly_ fixed :)
I'll leave this issue open as the primary issue for the right-click menu bug.
For the other tab favicon issues, please post any new issues you find on this thread: #478
